### PR TITLE
AEM app build caching

### PR DESCRIPTION
**Pull request by krystian-panek-wttech**
alternatively I checked other options

- https://github.com/apache/maven-build-cache-extension - the problem is not yet released Maven
- bash scripting - the problem with cross OS checksum calculation for dirs

finally I implemented the following:

```shell
#!/usr/bin/env sh

step "build AEM application"
aem app build \
  --command "mvn clean package" \
  --sources "pom.xml,all,core,ui.apps,ui.apps.structure,ui.config,ui.content,ui.frontend,ui.tests" \
  --file "all/target/*.all-*.zip"
clc

step "deploy AEM application"
aem package deploy --file "all/target/*.all-*.zip"
clc

```

the `sources` parameter cannot be avoided because one of root's child node contains AEM instances `aem/home`, other one `it.tests`, and the projects could have potentially more such "extra" dirs which are not connected to the building AEM app / artifacts. So I decided that to be able to leverage the caching it is required to simply list the source directories. For AEM archetype-based projects it looks quite simple.

image

image

image

image

@tomasz-sobczyk-wttech WDYT? :) 

take a look here - https://github.com/wttech/aemc/pull/32/files#r1068726035

### CODE REVIEW

File: pkg/common/filex/filex.go [hunk 1]
```diff
@@ -70,3 +75,53 @@ func Copy(sourcePath, destinationPath string) error {
...
+	defer func() {
+		_ = pathx.DeleteIfExists(dirTarFile)
+	}()
+	sort.Strings(tarFiles)
+	/* TODO remove it
+	for _, f := range tarFiles {
+		println(f)
+	}
+	*/
+	// TODO calculate hashes separately; dump to file and compare what is changing
```
Comment (krystian-panek-wttech) [inline pkg/common/filex/filex.go:118]: inspired by https://gist.github.com/jonathongardner/d0cdc191f6aec3251fd06c0dc03c3de1

File: pkg/facade.go [hunk 2]
```diff
@@ -44,6 +49,55 @@ func (a *Aem) InstanceManager() *InstanceManager {
 }
 
 func (a *Aem) Configure(config *cfg.Config) {
+	a.baseOpts.Configure(config)
 	a.javaOpts.Configure(config)
 	a.instanceManager.Configure(config)
 }
+
+func (a *Aem) Build(command string, outputFile string, inputPaths, inputExcludes []string) (bool, error) {
```
Comment (krystian-panek-wttech) [inline pkg/facade.go:57]: clue of change / @tomasz-sobczyk-wttech

File: pkg/base/base.go [hunk 3]
```diff
@@ -1,11 +1,56 @@
...
 type Opts struct {
-	TmpDir string
+	TmpDir           string
+	ChecksumExcludes []string
 }
 
 func NewOpts() *Opts {
 	return &Opts{
 		TmpDir: "aem/home/tmp",
+		ChecksumExcludes: []string{
```
Comment (krystian-panek-wttech) [inline pkg/base/base.go:15]: @tomasz-sobczyk-wttech if some entry looks wrong here let me know

File: pkg/base/base.go [hunk 4]
```diff
@@ -1,11 +1,56 @@
...
+			// AEM files
+			"**/aem/home/**",
+
+			// meta files
+			"**/.*",
+			"**/.*/**",
+			"!.content.xml",
```
Comment (krystian-panek-wttech) [inline pkg/base/base.go:22]: does AEM have any more files with such strange name? I mean dot in the beginning?

File: pkg/facade.go [hunk 5]
```diff
@@ -44,6 +49,55 @@ func (a *Aem) InstanceManager() *InstanceManager {
...
+	outputFileExists, err := pathx.ExistsStrict(outputFile)
+	if err != nil {
+		return false, err
+	}
+	checksumFile := outputFile + ".build"
```
Comment (krystian-panek-wttech) [inline pkg/facade.go:62]: image

**Comment by tomasz-sobczyk-wttech** [comment]
I have a feeling that this shouldnt be part of aem composr but rather the task runner that is "around it"

**Comment by krystian-panek-wttech** [comment]
I agree, if Maven or Typescript-based task runner around it will have such feature I will be ok to deprecate this one.
For now its beneficial for sure.

**Pull request merged**